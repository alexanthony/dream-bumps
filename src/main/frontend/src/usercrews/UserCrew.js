import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import {Card, CardTitle, CardText} from 'material-ui/Card';
import { Table, TableBody } from 'material-ui/Table';

import { getUserCrewMembers, getUserCrewRanking } from '../rootReducer';
import { fetchUserCrewMembers, fetchUserCrewRanking, setBuyMember } from './actions';
import UserCrewMemberRow from './UserCrewMemberRow';

/**
 * CurrentUserCrew
 */
class UserCrew extends Component {
  componentDidMount() {
    if (this.props.userId) {
      this.loadCrew(this.props);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.userId && this.props.userId !== nextProps.userId) {
      this.loadCrew(nextProps);
    }
  }

  loadCrew(props) {
    this.props.fetchUserCrewMembers(props.userId, this.props.sex);
    this.props.fetchUserCrewRanking(props.userId, this.props.sex);
  }

  render() {
    return (
      <Card>
        <CardTitle title={this.props.header} />
        <CardText>
          <Table
            selectable={false}
          >
            <TableBody
              displayRowCheckbox={false}
            >
              {this.props.crewMembers.map(member => (
                <UserCrewMemberRow
                  key={member.id}
                  member={member}
                  canBuySell={this.props.canBuySell}
                  onClickBuy={e => this.props.onClickBuy(member.id, this.props.sex)}
                />))
              }
            </TableBody>
          </Table>
        </CardText>
        {this.props.showCashAndValue &&
          <CardText>
            <dl className="dl-horizontal">
              <dt>Cash</dt>
              <dd>{this.props.crewRanking.cash ? this.props.crewRanking.cash.toFixed(2) : ''}</dd>
              <dt>Crew Value</dt>
              <dd>{this.props.crewRanking.value ? this.props.crewRanking.value.toFixed(2): ''}</dd>
            </dl>
          </CardText>
        }
      </Card>
    );
  }
}

UserCrew.propTypes = {
  sex: PropTypes.string,
  userId: PropTypes.number,
  crewMembers: PropTypes.array,
  fetchUserCrewMembers: PropTypes.func,
  fetchUserCrewRanking: PropTypes.func,
  onClickBuy: PropTypes.func,
  crewRanking: PropTypes.object,
  canBuySell: PropTypes.bool,
  header: PropTypes.string,
  showCashAndValue: PropTypes.bool
};

export const mapStateToProps = (state, ownProps) => ({
  crewMembers : getUserCrewMembers(state)(ownProps.userId, ownProps.sex),
  crewRanking : getUserCrewRanking(state)(ownProps.userId, ownProps.sex) || {}
});

export {UserCrew};

export default connect(mapStateToProps, { fetchUserCrewMembers, fetchUserCrewRanking, onClickBuy: setBuyMember })(UserCrew);
