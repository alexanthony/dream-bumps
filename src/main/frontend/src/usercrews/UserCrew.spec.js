import React from 'react';
import { shallow } from 'enzyme';
import { UserCrew } from './UserCrew';
import toJson from 'enzyme-to-json';

test('Renders properly', () => {
  const fetchUserCrewMembers = jest.fn();
  const fetchUserCrewRanking = jest.fn();
  const onClickBuy = jest.fn();
  const crewMembers = [{id: 3}, {id: 7}];
  const component = shallow(
    <UserCrew
      sex="male"
      userId={17}
      crewMembers={crewMembers}
      fetchUserCrewMembers={fetchUserCrewMembers}
      fetchUserCrewRanking={fetchUserCrewRanking}
      onClickBuy={onClickBuy}
      crewRanking={{cash: 123.4, value: 1067.12}}
      header="UserCrew"
      showCashAndValue
      canBuySell
    />
  );
  expect(component).toBeDefined();
  expect(toJson(component)).toMatchSnapshot()
  // We don't actually mount, so these don't get called...
  // expect(fetchUsers).toHaveBeenCalled();
  // expect(fetchUserCrewRankings).toHaveBeenCalled()
});

test('Renders properly - not buy/sell', () => {
  const fetchUserCrewMembers = jest.fn();
  const fetchUserCrewRanking = jest.fn();
  const crewMembers = [{id: 3}, {id: 7}];
  const component = shallow(
    <UserCrew
      sex="female"
      userId={17}
      crewMembers={crewMembers}
      fetchUserCrewMembers={fetchUserCrewMembers}
      fetchUserCrewRanking={fetchUserCrewRanking}
      crewRanking={{cash: 123.4, value: 1067.12}}
      header="Other Header"
      showCashAndValue={false}
      canBuySell={false}
    />
  );
  expect(component).toBeDefined();
  expect(toJson(component)).toMatchSnapshot()
  // We don't actually mount, so these don't get called...
  // expect(fetchUsers).toHaveBeenCalled();
  // expect(fetchUserCrewRankings).toHaveBeenCalled()
});
