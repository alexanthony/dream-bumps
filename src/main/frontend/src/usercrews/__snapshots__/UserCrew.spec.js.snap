exports[`test Renders properly - not buy/sell 1`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}>
  <CardTitle
    title="Other Header" />
  <CardText>
    <Table
      allRowsSelected={false}
      fixedFooter={true}
      fixedHeader={true}
      height="inherit"
      multiSelectable={false}
      selectable={false}>
      <TableBody
        allRowsSelected={false}
        deselectOnClickaway={true}
        displayRowCheckbox={false}
        multiSelectable={false}
        preScanRows={true}
        selectable={true}
        style={Object {}}>
        <Connect(UserCrewMemberRow)
          canBuySell={false}
          member={
            Object {
              "id": 3,
            }
          }
          onClickBuy={[Function]} />
        <Connect(UserCrewMemberRow)
          canBuySell={false}
          member={
            Object {
              "id": 7,
            }
          }
          onClickBuy={[Function]} />
      </TableBody>
    </Table>
  </CardText>
</Card>
`;

exports[`test Renders properly 1`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}>
  <CardTitle
    title="UserCrew" />
  <CardText>
    <Table
      allRowsSelected={false}
      fixedFooter={true}
      fixedHeader={true}
      height="inherit"
      multiSelectable={false}
      selectable={false}>
      <TableBody
        allRowsSelected={false}
        deselectOnClickaway={true}
        displayRowCheckbox={false}
        multiSelectable={false}
        preScanRows={true}
        selectable={true}
        style={Object {}}>
        <Connect(UserCrewMemberRow)
          canBuySell={true}
          member={
            Object {
              "id": 3,
            }
          }
          onClickBuy={[Function]} />
        <Connect(UserCrewMemberRow)
          canBuySell={true}
          member={
            Object {
              "id": 7,
            }
          }
          onClickBuy={[Function]} />
      </TableBody>
    </Table>
  </CardText>
  <CardText>
    <dl
      className="dl-horizontal">
      <dt>
        Cash
      </dt>
      <dd>
        123.40
      </dd>
      <dt>
        Crew Value
      </dt>
      <dd>
        1067.12
      </dd>
    </dl>
  </CardText>
</Card>
`;
